### PR TITLE
[DataGrid] Improve performance of width resizing

### LINKED ISSUE #799: [DataGrid] Poor performance when container is resized
Animations are stuttering when using Datagrid.
 
- [ ] The issue is present in the latest release.
version:  "@material-ui/core": "^4.11.2",
- [x] I have searched the [issues](https://github.com/mui-org/material-ui/issues) of this repository and believe that this is not a duplicate.

## Current Behavior 😯

Opening/Closing navigation Drawer stutters.
Resizing window stutters.

## Expected Behavior 🤔

Animations expected to be smooth when there are less than 100 data.

## Steps to Reproduce 🕹
All codes are taken from sample codes.

https://codesandbox.io/s/material-demo-forked-hb1k4?file=/demo.tsx

Steps:

1.Opening navigation Drawer stutters when using  Posts  component.
2.Change Posts to  EnhancedTable in demo.tsx
3.Stutter disappers and works as expected .

## Context 🔦

Trying to achive smooth animations when using DataGrid component. 

## Your Environment 🌎
 
  
```
 System:
    OS: Windows 10 10.0.19042
  Binaries:
    Node: 12.16.2 - C:\Program Files\nodejs\node.EXE
    Yarn: Not Found
    npm: 6.14.4 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Chrome: 87.0.4280.88
    Edge: Spartan (44.19041.423.0), Chromium (87.0.664.66) (Using Edge browser)
  npmPackages:
    @material-ui/core: ^4.11.2 => 4.11.2
    @material-ui/data-grid: ^4.0.0-alpha.14 => 4.0.0-alpha.14
    @material-ui/icons: ^4.11.2 => 4.11.2
    @material-ui/lab: ^4.0.0-alpha.57 => 4.0.0-alpha.57
    @material-ui/styles:  4.11.2
    @material-ui/system:  4.11.2
    @material-ui/types:  5.1.0
    @material-ui/utils:  4.11.2
    @types/react: ^16.14.2 => 16.14.2
    react: ^17.0.1 => 17.0.1
    react-dom: ^17.0.1 => 17.0.1
    typescript: ^4.1.2 => 4.1.2
```
</details>

### PR DESCRIPTION
**Pull request by dtassone**
fix #799 

Preview: https://deploy-preview-1686--material-ui-x.netlify.app/components/data-grid/

### CODE REVIEW

File: packages/grid/_modules_/grid/hooks/utils/useResizeContainer.ts [hunk 1]
```diff
@@ -8,6 +9,15 @@ export function useResizeContainer(apiRef): (size: ElementSize) => void {
   const gridLogger = useLogger('useResizeContainer');
   const { autoHeight } = useGridSelector(apiRef, optionsSelector);
 
+  const debounceResize = React.useMemo(
+    () =>
+      debounce((size: ElementSize) => {
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/hooks/utils/useResizeContainer.ts:14]: We can't use the debounce API exactly like this. It schedules a timeout that can fire after the component is unmounted. We should call clear(). For instance: https://github.com/mui-org/material-ui/blob/33850e52ca32787fbc02b64d3225bac2d451612c/packages/material-ui/src/Tabs/ScrollbarSize.js#L41

or even better. Here is the line of code you had removed: https://github.com/mui-org/material-ui-x/pull/544/files#diff-46fae3f9bc98017095df24cd6b014aa02cf1992249906d4d5ef92103857c11b4L46

Comment (dtassone) [reply packages/grid/_modules_/grid/hooks/utils/useResizeContainer.ts:14]: wow nice digging 😆 
Yeah I add the useEffect to clear the timeout 👍

File: packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx [hunk 2]
```diff
@@ -221,7 +222,7 @@ describe('<DataGrid /> - Layout & Warnings', () => {
           </div>,
         );
         const firstHeight = getByRole('grid').clientHeight;
-        await sleep(10);
+        clock.tick(10);
```
Comment (m4theushw) [inline packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx:225]: The `async` on this test case can be removed.

File: packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx [hunk 3]
```diff
@@ -59,7 +69,8 @@ describe('<DataGrid /> - Layout & Warnings', () => {
       expect(rect.width).to.equal(300 - 2);
 
       setProps({ width: 400 });
-      await raf(); // wait for the AutoSize's dimension detection logic
+      clock.tick(60);
```
Comment (m4theushw) [inline packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx:72]: The `async` on this test case can be removed.

File: packages/grid/_modules_/grid/hooks/utils/useResizeContainer.ts [hunk 4]
```diff
@@ -8,6 +9,15 @@ export function useResizeContainer(apiRef): (size: ElementSize) => void {
...
+        gridLogger.info('resized...', size);
+        apiRef.current.resize();
+      }, 60),
```
Comment (DanailH) [inline packages/grid/_modules_/grid/hooks/utils/useResizeContainer.ts:17]: Is there a reason why it is 60? I mean it looks like an arbitrary number.

Comment (dtassone) [reply packages/grid/_modules_/grid/hooks/utils/useResizeContainer.ts:17]: yes it's a bit arbitrary. We had a chat about it https://github.com/mui-org/material-ui-x/issues/799#issuecomment-753311355

To me >100 seems a bit high so I thought 60 was good. Happy to change?

File: packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx [hunk 5]
```diff
@@ -41,6 +51,7 @@ describe('<DataGrid /> - Layout & Warnings', () => {
     });
 
     it('should resize the width of the columns', async () => {
+      clock.restore();
```
Comment (m4theushw) [inline packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx:54]: Why to restore again?  It's already called in `afterEach`.

Comment (dtassone) [reply packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx:54]: because this one needs to wait for raf due to autosizer.

Comment (dtassone) [reply packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx:54]: I moved the clock one scope above.

**Review by m4theushw — APPROVED** [review]

**Comment by ficu88** [comment]
Hello, everyone, im still having the performance resizing problem with the following libraries:

```
"@mui/material": "^5.2.3",
"@mui/x-data-grid": "^5.3.0",
```

Like this guy's problem:
[https://stackoverflow.com/questions/70482864/material-ui-datagrid-performance-issues](https://stackoverflow.com/questions/70482864/material-ui-datagrid-performance-issues)

Thank you

**Comment by VladBrok** [comment]
I have a similar performace issue too (like this: https://github.com/mui/mui-x/issues/799). When I resize the browser window, CPU consumtion goes from 0.12% to 10%-20%.

**Pull request merged**